src/interleave-generator.mjs

```diff
@@ -0,0 +1,42 @@
+import { ensureIterable } from './internal/iterable'
+import map from './map'
+
+class InterleaveBuffer {
+  constructor (iterator) {
+    this._iterator = iterator
+    this._next = this._iterator.next()
+  }
+
+  take () {
+    const next = this._next
+    if (next.done) {
+      return undefined
+    }
+
+    this._next = this._iterator.next()
+    return next.value
+  }
+
+  canTake () {
+    return this._next.done ? null : this
+  }
+}
+
+export default function interleaveGenerator (generatorFn) {
+  return (...iterables) => {
+    return (function * () {
+      const iterators = map(iterable => ensureIterable(iterable)[Symbol.iterator](), iterables)
+      const buffers = Array.from(map(iterator => new InterleaveBuffer(iterator), iterators))
+      const canTakeAny = () => buffers.find(buffer => buffer.canTake())
```

Comment:
I am not sure if `canTakeAny` is an appropriate name since it implies returning `boolean` ("can \<something\>?" is a yes/no question)

Reply:
I struggled with that same thought for a while too. I decided to name it based on its primary intended usage. I think what I aught to do is when it is being used in a different way my example code should bind it to a different name, like `findTakableBuffer`. I have the one example like that at the bottom of the docs for the async version.